File: datacenter_cooling.py
# Author: Chernoy Viacheslav
# email: vchenoy@gmail.com
#
# The program solves "Datacenter Cooling" Problem presented by Quora
# http://www.quora.com/challenges

# reads and returns 0/1/2/3 matrix from the standard-input
def read_input():
    datacenter = []
    w, h = [int(x) for x in raw_input().split()]
    for i in xrange(h):
        l = [int(x) for x in raw_input().split()]
        assert(len(l) == w)

        datacenter.append(l)

    return datacenter

# datacenter is 0/1/2/3-matrix
# creates and returns the graph representation of datacenter
# also returns source and destination verticies
def build_graph(datacenter):
    # converts the pair (i, j) to a vertex
    def vert(i, j):
        return i * w + j
    
    OWN       = 0
    START     = 2
    END       = 3
    DONOT_OWN = 1

    h = len(datacenter)
    w = len(datacenter[0])
    # the number of verticies
    n = w * h
    # create empty graph
    graph = [[] for k in xrange(n)]
    # four directions up, down, left, right
    D = [(-1,0), (1,0), (0,-1), (0,1)]
    for i in xrange(h):
        for j in xrange(w):
            # the current vertex v
            v = vert(i, j)
            # adjacency list of v
            adj = []
            if datacenter[i][j] in [OWN, START]:
                # find all verticies adjacent to v
                for d in D:
                    i1 = i + d[0]
                    j1 = j + d[1]
                    if (i1 >= 0) and (i1 < h) and (j1 >= 0) and (j1 < w) and (datacenter[i1][j1] != DONOT_OWN):
                        u = vert(i1, j1)
                        adj.append(u)

            if datacenter[i][j] == START:
                src = v
            elif datacenter[i][j] == END:
                dst = v 
         
            graph[v] = adj

    return src, dst, graph

# graph is a graph
# L is the list of vertices to check
# u is a destination vertex
# checks whether the vertex u is reachable from all the vertices of L
def connected(graph, L, u):
    n = len(graph)
    L = L[:]
    # C[v] == 0 means it is not connected
    # for every k >= 2, {v|C[v] == k} are connected to v = L[1-k] (and between each other)
    C = [0] * n
    C[u] = 1
    mark = 2

    while len(L) > 0:
        v = L.pop()
        # check whether v can reach u
        if C[v] == 0:
            # R is a list of verticies reachable form v
            R = [v]
            while True:
                if len(R) == 0:
                    # v cannot reach u
                    return False

                # x is reachable from v
                x = R.pop()
                # check its adjacent ones
                for y in graph[x]:
                    if C[y] == 0:
                        # y is reachable from v
                        C[y] = mark 
                        R.append(y)
                    elif C[y] < mark:
                        # u is reachable from y,
                        # so we are done with v
                        break

                if C[y] < mark:
                    break

            mark += 1

    return True    

# v is the current vertex
# l is the number of steps remained to do
# backtracking, computes the number of possible paths of length l from v to dst
def search(v, l):
    global dst, count, visited, graph

    Y = [u for u in graph[v] if not visited[u]]
    X = []
    # remove all the edges to v
    for u in Y:
        if v in graph[u]:
            if len(graph[u]) <= 1:
                # u is isolated, rollback
                for w in X:
                    graph[w].append(v)

                return

            # updating the graph
            graph[u].remove(v)
            # save, for later restore it
            X.append(u)

    visited[v] = True
    # check that dst is reachable from all the vertices of X 
    if (len(X) == 0) or connected(graph, X, dst):
        # one step less is remained to dst
        l -= 1
        # go over all possible steps that can be done
        # i.e., check every unvisited adjacent vertex
        for u in Y:
            if (l == 0) and (u == dst):
                # we reached dst and passed all verticies
                count += 1
                #if count % 1000 == 0:
                #    print count
            elif (l > 0) and (u != dst):
                # not yet reached dst, let's check the step
                search(u, l)

    visited[v] = False

    # restore the graph
    for w in X:
        graph[w].append(v)



datacenter = read_input()
src, dst, graph = build_graph(datacenter)
n = len(datacenter) * len(datacenter[0])
# compute the number of verticies (the path length) we have to pass to reach dst
length = len([v for v in xrange(n) if len(graph[v]) > 0])
visited = [False] * n
# the number of pathes
count = 0
search(src, length)
print count

File: datacenter_cooling.go
// Author: Chernoy Viacheslav
// email: vchenoy@gmail.com
//
// The program solves "Datacenter Cooling" Problem presented by Quora
// http://www.quora.com/challenges

package main

import (
    "fmt"
)

type (
    // the 0/1/2/3-matrix type
    grid_t [][] int

    vertex_t int

    // the graph represented by adjacency list
    // the list type (maximum 4 (four) adjacent verticies exist
    graph_t [][] vertex_t
)

var (
    count int
    visited [] bool
    graph graph_t
    src, dst vertex_t
    R [] vertex_t
)


// list sub-system
func index(list [] vertex_t, x vertex_t) int {
    i := len(list)-1
    for (i >= 0 && list[i] != x) {
        i -= 1
    }

    return i
}

// reads 0/1/2/3 matrix from the standard-input
func read_input(datacenter *grid_t) {
    var (
        w, h int
    )

    _, _ = fmt.Scanf("%d %d", &w, &h)
    *datacenter = make(grid_t, h)
    for i := range *datacenter {
        (*datacenter)[i] = make([]int, w)
        for j := range (*datacenter)[i] {
            _, _ = fmt.Scanf("%d", &(*datacenter)[i][j])
        }
    }
}

// converts the pair (i, j) to a vertex
func vert(w int, i, j int) vertex_t {
    return vertex_t(i * w + j)
}

// datacenter is the 0/1/2/3-grid
// graph is adjacency list contructed from the matrix datacenter
// builds the graph representation of datacenter
// also returns source and destination vertices
func build_graph(datacenter grid_t, graph *graph_t, src, dst *vertex_t) {
    // four directions up, down, left, right
    const (
        // the room types
        OWN       = 0
        START     = 2
        END       = 3
        DONOT_OWN = 1
    )

    Dh := [4]int{-1, 1, 0, 0}
    Dw := [4]int{0, 0, -1, 1}

    // create empty graph
    n := len(datacenter) * len(datacenter[0])
    *graph = make(graph_t, 0, n)

    for i := range datacenter {
        for j := range datacenter[i] {
            // the current vertex v
            v := vert(len(datacenter[i]), i, j)
            // adjacency list of v
            adj := make([]vertex_t, 0, 4)
            if datacenter[i][j] == OWN || datacenter[i][j] == START {
                // find all verticies adjacent to v
                for k := range Dh {
                    i1 := i + Dh[k]
                    j1 := j + Dw[k]
                    if (i1 >= 0) && (i1 < len(datacenter)) && (j1 >= 0) && (j1 < len(datacenter[i])) && (datacenter[i1][j1] != DONOT_OWN) {
                        u := vert(len(datacenter[i]), i1, j1)
                        adj = append(adj, u)
                    }
                }
            }
            *graph = append(*graph, adj)
            if datacenter[i][j] == START {
                *src = v
            } else if datacenter[i][j] == END {
                *dst = v
            }
        }
    }
}

// graph represented by adjacency list
// L is the list of vertices to check
// u is a destination vertex
// checks whether the vertex u is reachable from all the vertices of L
func connected(graph graph_t, L [] vertex_t, u vertex_t) bool {
    // global: R

    // C[v] == 0 means it is not connected
    // for every k >= 2, {v|C[v] == k} are connected to v = L[1-k] (and between each other)
    C := make([]int, len(graph))
    C[u] = 1
    mark := 2
    for _, v := range L {
        // check whether v can reach u
        if C[v] == 0 {
            R = R[:1]
            R[0] = v
            lb:
            for {
                if len(R) == 0 {
                    // v cannot reach u
                    return false
                }
                // x is reachable from v
                x := R[len(R)-1]
                R = R[:len(R)-1]
                // check its adjacent ones
                for _, y := range graph[x] {
                    if C[y] == 0 {
                        // y is reachable from v
                        C[y] = mark
                        R = append(R, y)
                    } else if C[y] < mark {
                        // u is reachable from y,
                        // so we are done with v
                        break lb 
                    }
                }
            }
            mark += 1
        }
    }
    return true
}

// v is the current vertex
// l is the number of steps remained to do
// backtracking, computes the number of possible paths of length l from v to dst
func search(v vertex_t, l int) {
    // global: dst, count, visited, graph

    X := make([]vertex_t, 0, 4)
    Y := make([]vertex_t, 0, 4)
    for _, u := range graph[v] {
        if !visited[u] {
            Y = append(Y, u)
        }
    }
    // remove all the edges to v
    for _, u := range Y {
        if index(graph[u], v) >= 0 {
            if len(graph[u]) <= 1 {
                // u is isolated, rollback
                for _, u := range X {
                    graph[u] = append(graph[u], v)
                }
                return
            }
            // updating the graph
            i := index(graph[u], v)
            graph[u][i] = graph[u][len(graph[u])-1]
            graph[u] = graph[u][:len(graph[u])-1]

            // save, for later restore it
            X = append(X, u)
        }
    }
    visited[v] = true
    // check that dst is reachable from all the vertices of X 
    if len(X) == 0 || connected(graph, X, dst) {
        // one step less is remained to dst
        l -= 1
        // go over all possible steps that can be done
        // i.e., check every unvisited adjacent vertex
        for _, u := range Y {
            if l == 0 && u == dst {
                // we reached dst and passed all verticies
                count += 1
                //if (count % 1000 == 0) {
                //    fmt.Printf("%d\n", count)
                //}
            } else if l > 0 && u != dst {
                // not yet reached dst, let's check the step
                search(u, l)
            }
        }
    }
    visited[v] = false

    // restore the graph
    for _, u := range X {
        graph[u] = append(graph[u], v)
    }
}

func main() {
    var (
        datacenter grid_t
    )

    read_input(&datacenter)
    build_graph(datacenter, &graph, &src, &dst)
    // compute the number of verticies (the path length) we have to pass to reach dst
    n := len(graph)
    R = make([]vertex_t, 0, n)
    length := 0
    for _, adj := range graph {
        if len(adj) > 0 {
            length += 1
        }
    }
    visited = make([]bool, n)
    // the number of pathes
    count = 0
    search(src, length)
    fmt.Printf("%d\n", count)
}

